polars/polars-io/src/csv/read.rs

```diff
@@ -496,35 +502,80 @@ where
         #[cfg(feature = "dtype-categorical")]
         let mut _cat_lock = None;
 
-        let mut df = if let Some(schema) = schema_overwrite {
-            let (schema, to_cast, _has_cat) = self.prepare_schema_overwrite(schema);
+        let (mut df, hack_flag) = match (schema_overwrite, dtype_overwrite) {
+            (Some(schema), _) => {
+                let (schema, to_cast, _has_cat) = self.prepare_schema_overwrite(schema);
+
+                #[cfg(feature = "dtype-categorical")]
+                if _has_cat {
+                    _cat_lock = Some(polars_core::IUseStringCache::new())
+                }
 
-            #[cfg(feature = "dtype-categorical")]
-            if _has_cat {
-                _cat_lock = Some(polars_core::IUseStringCache::new())
+                let mut csv_reader = self.core_reader(Some(&schema), to_cast)?;
+                (csv_reader.as_df()?, false)
             }
+            (None, Some(dtypes)) => {
+                // we need to get the column names from the header
+                let reader_bytes = get_reader_bytes(&mut self.reader)?;
```

Comment:
This means we call this one twice on a csv read. This can be a full disk read in the case of compressed csv's so if we call this we should keep the reader_bytes and initialize the core reader with it.

Reply:
Yup, and I agree that this is far from a decent approach. I did not want to get too into the weeds with the whole header-parsing logic though as that seemed well beyond my current skill level in terms of code I would be touching for this small problem. I would very much welcome some suggestions to remove this (silly) two-step solution.

Reply:
I think we should could add an `optional` reader_bytes argument to the core reader.

Then you pass the reader_bytes you just created and then let the core reader continue doing its work.